graphql/end_to_end_test.go

```diff
@@ -201,3 +201,104 @@ func TestEndToEndAwaitAndCache(t *testing.T) {
 		t.Errorf("expected 4 total calls to slow, got %d", calls)
 	}
 }
+
+func TestArgumentOptionality(t *testing.T) {
+	schema := schemabuilder.NewSchema()
+	query := schema.Query()
+
+	query.FieldFunc("optional", func(args struct{ X *int64 }) int64 {
+		if args.X != nil {
+			return *args.X
+		}
+		return -1
+	})
+
+	_ = schema.Mutation()
+
+	builtSchema := schema.MustBuild()
+
+	// argument is optional and not passed in
+	q := graphql.MustParse(`
+		query getOptional($testArg: int64) {
+			optional(x: $testArg)
+		}`, map[string]interface{}{})
+
+	if err := graphql.PrepareQuery(builtSchema.Query, q.SelectionSet); err != nil {
+		t.Error(err)
+	}
+
+	e := graphql.Executor{}
+	result, err := e.Execute(context.Background(), builtSchema.Query, nil, q)
+	if err != nil {
+		t.Error(err)
+	}
+
+	if !reflect.DeepEqual(internal.AsJSON(result), internal.ParseJSON(`
+		{"optional": -1}`)) {
+		t.Error(internal.AsJSON(result))
+	}
+
+	testVariables := map[string]interface{}{
+		"testArg": float64(5),
+	}
+
+	// arg is optional and passed in
+	q = graphql.MustParse(`
+		query getOptional($testArg: int64) {
+			optional(x: $testArg)
+		}`, testVariables)
+
+	if err := graphql.PrepareQuery(builtSchema.Query, q.SelectionSet); err != nil {
+		t.Error(err)
+	}
+
+	result, err = e.Execute(context.Background(), builtSchema.Query, nil, q)
+	if err != nil {
+		t.Error(err)
+	}
+
+	if !reflect.DeepEqual(internal.AsJSON(result), internal.ParseJSON(`
+		{"optional": 5}`)) {
+		t.Error(internal.AsJSON(result))
+	}
+
+	// arg is mandatory and not passed in
+	q = graphql.MustParse(`
+		query getOptional($testArg: int64!) {
+			optional(x: $testArg)
+		}`, map[string]interface{}{})
+
+	if err := graphql.PrepareQuery(builtSchema.Query, q.SelectionSet); err != nil {
+		t.Error(err)
+	}
+
+	result, err = e.Execute(context.Background(), builtSchema.Query, nil, q)
+	if err != nil {
+		t.Error(err)
+	}
+
+	if !reflect.DeepEqual(internal.AsJSON(result), internal.ParseJSON(`
+		{"optional": -1}`)) {
```

Comment:
I think I expect an error here, since we passed in nothing, but the variable is required

Reply:
Oh you're right. I think I need to make a "mandatory" field that has a matching go function definition

Reply:
(i.e. even though none of the users of the argument require it, the query itself should be requiring it.)